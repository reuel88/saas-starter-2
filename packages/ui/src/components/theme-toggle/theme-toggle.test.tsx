import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeToggle } from "./theme-toggle";
import { vi } from "vitest";
import * as nextTheme from "next-themes";

vi.mock("next-themes", async (importOriginal) => {
  return {
    ...(await importOriginal<typeof import("next-themes")>()),
    useTheme: () => vi.fn(() => {}),
  };
});

describe("ThemeToggle", () => {
  it("should render a dropdown menu with a theme toggle button", () => {
    // Arrange

    // Act
    render(<ThemeToggle />);
    const dropdownMenu = screen.getByRole("button", { name: "Toggle theme" });

    // Assert
    expect(dropdownMenu).toBeInTheDocument();
  });

  it("should show a dropdown menu when the theme toggle button is clicked", async () => {
    // Arrange
    const { baseElement } = render(<ThemeToggle />);

    // Act
    const dropdownMenuButton = screen.getByRole("button", {
      name: "Toggle theme",
    });
    await userEvent.click(dropdownMenuButton);
    const dropdownMenuContent = screen.getByRole("menu");
    const lightOption = screen.getByRole("menuitem", { name: "Light" });
    const darkOption = screen.getByRole("menuitem", { name: "Dark" });
    const systemOption = screen.getByRole("menuitem", { name: "System" });

    // Assert
    expect(dropdownMenuContent).toBeInTheDocument();
    expect(lightOption).toBeInTheDocument();
    expect(darkOption).toBeInTheDocument();
    expect(systemOption).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it("should call setTheme with 'light' when light option is clicked", async () => {
    // Arrange
    const setThemeMock = vi.fn();
    vi.spyOn(nextTheme, "useTheme").mockReturnValue({
      setTheme: setThemeMock,
      themes: [],
    });

    // Act
    render(<ThemeToggle />);
    const dropdownMenuButton = screen.getByRole("button", {
      name: "Toggle theme",
    });
    await userEvent.click(dropdownMenuButton);
    const option = screen.getByRole("menuitem", { name: "Light" });
    await userEvent.click(option);

    // Assert
    expect(setThemeMock).toHaveBeenCalledWith("light");
  });

  it("should call setTheme with 'dark' when dark option is clicked", async () => {
    // Arrange
    const setThemeMock = vi.fn();
    vi.spyOn(nextTheme, "useTheme").mockReturnValue({
      setTheme: setThemeMock,
      themes: [],
    });

    // Act
    render(<ThemeToggle />);
    const dropdownMenuButton = screen.getByRole("button", {
      name: "Toggle theme",
    });
    await userEvent.click(dropdownMenuButton);
    const option = screen.getByRole("menuitem", { name: "Dark" });
    await userEvent.click(option);

    // Assert
    expect(setThemeMock).toHaveBeenCalledWith("dark");
  });

  it("should call setTheme with 'system' when system option is clicked", async () => {
    // Arrange
    const setThemeMock = vi.fn();
    vi.spyOn(nextTheme, "useTheme").mockReturnValue({
      setTheme: setThemeMock,
      themes: [],
    });

    // Act
    render(<ThemeToggle />);
    const dropdownMenuButton = screen.getByRole("button", {
      name: "Toggle theme",
    });
    await userEvent.click(dropdownMenuButton);
    const option = screen.getByRole("menuitem", { name: "System" });
    await userEvent.click(option);

    // Assert
    expect(setThemeMock).toHaveBeenCalledWith("system");
  });
});
