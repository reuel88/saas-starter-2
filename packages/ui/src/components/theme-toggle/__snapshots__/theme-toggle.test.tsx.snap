// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeToggle > should show a dropdown menu when the theme toggle button is clicked 1`] = `
<body
  data-scroll-locked=""
  style="pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <button
      aria-controls="radix-:r3:"
      aria-expanded="true"
      aria-haspopup="menu"
      class="text-primary inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
      data-state="open"
      id="radix-:r2:"
      type="button"
    >
      <span
        class="sr-only"
      >
        Toggle theme
      </span>
      <svg
        class="lucide lucide-sun h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="4"
        />
        <path
          d="M12 2v2"
        />
        <path
          d="M12 20v2"
        />
        <path
          d="m4.93 4.93 1.41 1.41"
        />
        <path
          d="m17.66 17.66 1.41 1.41"
        />
        <path
          d="M2 12h2"
        />
        <path
          d="M20 12h2"
        />
        <path
          d="m6.34 17.66-1.41 1.41"
        />
        <path
          d="m19.07 4.93-1.41 1.41"
        />
      </svg>
      <svg
        class="lucide lucide-moon absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
        />
      </svg>
    </button>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    dir="ltr"
    style="position: fixed; left: 0px; top: 0px; transform: translate(0px, 4px); min-width: max-content; --radix-popper-available-width: 0px; --radix-popper-available-height: -4px; --radix-popper-anchor-width: 0px; --radix-popper-anchor-height: 0px; --radix-popper-transform-origin: 100% 0px;"
  >
    <div
      aria-labelledby="radix-:r2:"
      aria-orientation="vertical"
      class="z-50 min-w-[8rem] overflow-hidden rounded-md border bg-popover p-1 text-popover-foreground shadow-md data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
      data-align="end"
      data-orientation="vertical"
      data-radix-menu-content=""
      data-side="bottom"
      data-state="open"
      dir="ltr"
      id="radix-:r3:"
      role="menu"
      style="outline: none; --radix-dropdown-menu-content-transform-origin: var(--radix-popper-transform-origin); --radix-dropdown-menu-content-available-width: var(--radix-popper-available-width); --radix-dropdown-menu-content-available-height: var(--radix-popper-available-height); --radix-dropdown-menu-trigger-width: var(--radix-popper-anchor-width); --radix-dropdown-menu-trigger-height: var(--radix-popper-anchor-height); pointer-events: auto;"
      tabindex="-1"
    >
      <div
        class="relative flex cursor-default select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
        data-orientation="vertical"
        data-radix-collection-item=""
        role="menuitem"
        tabindex="-1"
      >
        Light
      </div>
      <div
        class="relative flex cursor-default select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
        data-orientation="vertical"
        data-radix-collection-item=""
        role="menuitem"
        tabindex="-1"
      >
        Dark
      </div>
      <div
        class="relative flex cursor-default select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
        data-orientation="vertical"
        data-radix-collection-item=""
        role="menuitem"
        tabindex="-1"
      >
        System
      </div>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
